scripts/generate_example_ttl_files.py

```diff
@@ -0,0 +1,120 @@
+    if actual_params != expected_params or defaults != (None, None, "N"):
+        raise RuntimeError("BNode.__new__ had unexpected signature, the following code might need to be updated")
+    # Override default _sn_gen function with our custom one.
+    rdflib.BNode.__new__.__defaults__ = (defaults[0], _custom_sn_gen, defaults[2])
+
+    # 2. second override:
+    # Mock "Date.today()" function to always give the same result
+    unittest.mock.patch.object(
+        abis_mapping.types.temporal.Date,
+        "today",
+        new=lambda: abis_mapping.types.temporal.Date(2020, 1, 1),
+    ).start()
+
+    # Finished overrides.
+
+    # Generate new ttl files.
+    for template_id, input_csv_file_path, output_ttl_file_path in examples:
+        print(f"Generating {output_ttl_file_path}...")
+
+        # set global var
+        TTL_FILE_PATH = str(output_ttl_file_path)
+
+        # Get Mapper
+        mapper = abis_mapping.get_mapper(template_id)
+        if mapper is None:
+            raise RuntimeError(f"Mapper not found for {template_id}")
+        # Map data
+        data = input_csv_file_path.read_bytes()
+        graphs = list(mapper().apply_mapping(data))
+        if len(graphs) != 1:
+            raise RuntimeError("apply_mapping did not produce exactly 1 graph")
+        # Write to output file
+        graphs[0].serialize(destination=output_ttl_file_path, format="turtle")
+
+    print("Done!")
+
+
+def _custom_sn_gen() -> str:
+    """Generate a BNode ID depending on the ttl file and the current code path.
+
+    Returns:
+        String to use for a BNode ID. Should be 32 hex digits to match rdflib default behavior.
+    """
+    # Generate ID with 32 hex digits (16 bytes) to match the normal behavior (which is uuid4().hex)
+    # First 12 bytes are from hash of ttl file path and the code path taken to create the BNode().
+    # Remaining 4 bytes are taken from a counter unique to the hash result.
+    # i.e. when the same hash occurs, they will be numbered 0,1,2,etc. at the end of the ID.
+    # This ensures each ID should be globally unique.
+    code_path_hash = hashlib.blake2b(digest_size=12, person=b"gen_b_node_id")
+    # hash ttl file path.
+    code_path_hash.update(TTL_FILE_PATH.encode("utf-8"))
+    # Hash functions in the abis_mapping module called to make this BNode()
+    for frame in inspect.stack():
+        # NOTE "abis_mapping" with underscore is the module name.
+        # NOT "abis-mapping" with a dash which is the repository name.
+        if "abis_mapping" in frame.filename:
+            code_path_hash.update(frame.function.encode("utf-8"))
+            # we don't include the frame.filename in the hash because this wil be
+            # different depending on where abis-mapping is checked out.
+    digest = code_path_hash.digest()
+    number = next(COUNTERS[digest]).to_bytes(4, "big")
+    return (digest + number).hex()  # concatenate bytes and return hex representation
```

Comment:
not sure this level of name creation is needed. I think all that is required is that the ID starts with a letter (in rdflib they use "N") but after that can use alphanumerics and a few select symbols. Why not just make the id a sequential number generator with a prefix of "L". A key thing to note is that [blank nodes don't resolve outside of a file](https://www.w3.org/TR/rdf12-concepts/#section-blank-nodes).

Reply:
"L" for Lincoln? :laughing: 

I didn't use a simple sequential number generator because then a mapping change which changes the number of blank nodes created, will change the IDs of all following B nodes.
The point of this (I admit complex) generation, is to try to have the B Node IDs change as little as possible, even when mapping is changed, so the git diff is reduced

Reply:
Pretty certain the "N" is for Nick. Solid reasoning for going this route.